### PR TITLE
additonal parser for aquasec trivy container vulnability scanner

### PR DESCRIPTION
**Pull request by tofuatjava**

### CODE REVIEW

**Comment by uhafner** [comment]
Can you please have a look at the analysis issues first?

**Comment by tofuatjava** [comment]
now it looks well, a 93,1% coverage should be good enough
until this is merged I will go one with the Warnings-ng implementation, I have still troubles with compiling it, due to dependency issues

File: src/main/java/edu/hm/hafner/analysis/parser/TrivyParser.java [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+    public Report parse(final ReaderFactory readerFactory) throws ParsingException, ParsingCanceledException {
+        final Report report = new Report();
+
+        try (Reader reader = readerFactory.create()) {
+            final JSONArray jsonReport = (JSONArray)new JSONTokener(reader).nextValue();
+
+            for (int i = 0; i < jsonReport.length(); i++) {
+                final JSONObject component = (JSONObject)jsonReport.get(i);
+                if (!component.isNull("Vulnerabilities")) {
+                    final JSONArray vulnatbilites = component.getJSONArray("Vulnerabilities");
```
Comment (uhafner) [inline src/main/java/edu/hm/hafner/analysis/parser/TrivyParser.java:48]:
```suggestion
                    final JSONArray vulnerabilities = component.getJSONArray("Vulnerabilities");
```

File: src/main/java/edu/hm/hafner/analysis/parser/TrivyParser.java [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+                    for (Object vulnatbility : vulnatbilites) {
```
Comment (uhafner) [inline src/main/java/edu/hm/hafner/analysis/parser/TrivyParser.java:49]:
```suggestion
                    for (Object vulnerability : vulnatbilites) {
```

File: src/main/java/edu/hm/hafner/analysis/parser/TrivyParser.java [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+                        report.add(convertToIssue((JSONObject)vulnatbility));
```
Comment (uhafner) [inline src/main/java/edu/hm/hafner/analysis/parser/TrivyParser.java:50]:
```suggestion
                        report.add(convertToIssue((JSONObject) vulnerability));
```

File: src/main/java/edu/hm/hafner/analysis/parser/TrivyParser.java [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+                    }
+                }
+            }
+        }
+        catch (IOException e) {
```
Comment (uhafner) [inline src/main/java/edu/hm/hafner/analysis/parser/TrivyParser.java:55]: I think this should be better:
```
catch (IOException | JSONException | ClassCastException e) {
```

File: src/main/java/edu/hm/hafner/analysis/parser/TrivyParser.java [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+            throw new ParsingException(e);
+        }
+
+        return report;
+    }
+
+    private Issue convertToIssue(final JSONObject vulnerability) {
+        return new IssueBuilder().setFileName(vulnerability.optString("PkgName", "?"))
```
Comment (uhafner) [inline src/main/java/edu/hm/hafner/analysis/parser/TrivyParser.java:63]: Undefined is `"-"` in all of my parsers.

File: src/test/java/edu/hm/hafner/analysis/parser/TrivyParserTest.java [hunk 6]
```diff
@@ -0,0 +1,31 @@
+package edu.hm.hafner.analysis.parser;
+
+import edu.hm.hafner.analysis.AbstractParserTest;
+import edu.hm.hafner.analysis.IssueParser;
+import edu.hm.hafner.analysis.Report;
+import edu.hm.hafner.analysis.Severity;
+import edu.hm.hafner.analysis.assertions.SoftAssertions;
+
```
Comment (uhafner) [inline src/test/java/edu/hm/hafner/analysis/parser/TrivyParserTest.java:8]: Please add a simple JavaDoc to the class, see other tests.

File: src/test/java/edu/hm/hafner/analysis/parser/TrivyParserTest.java [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+class TrivyParserTest extends AbstractParserTest {
+
+    TrivyParserTest() {
+        super("trivy_result.json");
+    }
+
+    @Override
+    protected void assertThatIssuesArePresent(final Report report, final SoftAssertions softly) {
+        softly.assertThat(report).hasSize(282);
```
Comment (uhafner) [inline src/test/java/edu/hm/hafner/analysis/parser/TrivyParserTest.java:17]: Can you please shorten the file, e.g. just 10 warnings?

**Review by uhafner — CHANGES_REQUESTED** [review]
Thanks for your pull request!

Does it make sense to check for the correct file ending? Then please add a test like
https://github.com/jenkinsci/analysis-model/blob/master/src/test/java/edu/hm/hafner/analysis/parser/HadoLintParserTest.java#L50

(In this example you can also see a test with corrupt input)

File: src/test/java/edu/hm/hafner/analysis/parser/TrivyParserTest.java (hunk 7 above)
Comment (tofuatjava) [reply src/test/java/edu/hm/hafner/analysis/parser/TrivyParserTest.java:17]: i just want to have a real world sample, but i will try to reduce...

File: src/test/java/edu/hm/hafner/analysis/parser/TrivyParserTest.java (hunk 6 above)
Comment (tofuatjava) [reply src/test/java/edu/hm/hafner/analysis/parser/TrivyParserTest.java:8]: I do think that a comment like "Tests the class xy" will improve anything, but its your code base.

Don't get me wrong comments are important but only on places where its needed to understand whats going on. I would rather suggest do reduce commenting but choose better names for methods variables mambers and so on...

File: src/main/java/edu/hm/hafner/analysis/parser/TrivyParser.java (hunk 4 above)
Comment (tofuatjava) [reply src/main/java/edu/hm/hafner/analysis/parser/TrivyParser.java:55]: normally I want to reduce the amount of code, especially when it is not essentially. In this case it is not important to know all these exceptions...
BTW: I'm also not a friend of butting RuntimeExceptions to the throws clause. Standard java rules in SonarLint will claim these approach.

**Comment by tofuatjava** [comment]
Thanks for the review.
In my opinion, checking for file endings in this case is unnecessary.
The report need to be done by command line and could be named without any convention. In the linux it is also not very common to give fileendings. But it is not clear what the better fileending would be ".json", ".jsn" or ".txt". there are no rules which specifies the fileending.
In worst case it makes no difference, because it is not garanteed that in a file with ".json" fileending a valid JSON is in it.

**Comment by tofuatjava** [comment]
I do not understand why I will break the API. Please advice...

**Comment by uhafner** [comment]
Can you please update the version in `pom.xml` to 9.5.0. Adding a new parser requires to increase the minor version. (Or merge with `master` that already contains that change.

**Review by uhafner — APPROVED** [review]

**Comment by tofuatjava** [comment]
I'm wondering the version should be already 9.5.0-SNAPSHOT
revision is set to 9.5.0 and changelist is set to -SNAPSHOT
target artifact is build as analysis-model-9.5.0-SNAPSHOT

maybe i need to increase once again after the rebase. next try with 9.6.0-SNAPSHOT

**Pull request merged**